test/Feature/RpcSpec.hs

```diff
@@ -274,21 +274,102 @@ spec actualPgVersion =
           { matchHeaders = [matchContentTypeJson] }
 
     context "proc argument types" $ do
-      it "accepts a variety of arguments" $
-        post "/rpc/varied_arguments"
-            [json| {
-              "double": 3.1,
-              "varchar": "hello",
-              "boolean": true,
-              "date": "20190101",
-              "money": 0,
-              "enum": "foo",
-              "integer": 43,
-              "json": {"some key": "some value"},
-              "jsonb": {"another key": [1, 2, "3"]}
-            } |]
+      -- different syntax for array needed for pg<10
+      when (actualPgVersion < pgVersion100) $
+        it "accepts a variety of arguments (Postgres < 10)" $
+          post "/rpc/varied_arguments"
+              [json| {
+                "double": 3.1,
+                "varchar": "hello",
+                "boolean": true,
+                "date": "20190101",
+                "money": 0,
+                "enum": "foo",
+                "arr": "{a,b,c}",
+                "integer": 43,
+                "json": {"some key": "some value"},
+                "jsonb": {"another key": [1, 2, "3"]}
+              } |]
+            `shouldRespondWith`
+              [json| {
+                "double": 3.1,
+                "varchar": "hello",
+                "boolean": true,
+                "date": "2019-01-01",
+                "money": "$0.00",
+                "enum": "foo",
+                "arr": ["a", "b", "c"],
+                "integer": 43,
+                "json": {"some key": "some value"},
+                "jsonb": {"another key": [1, 2, "3"]}
+              } |]
+              { matchHeaders = [matchContentTypeJson] }
+
+      when (actualPgVersion >= pgVersion100) $
+        it "accepts a variety of arguments (Postgres >= 10)" $
+          post "/rpc/varied_arguments"
+              [json| {
+                "double": 3.1,
+                "varchar": "hello",
+                "boolean": true,
+                "date": "20190101",
+                "money": 0,
+                "enum": "foo",
+                "arr": ["a", "b", "c"],
+                "integer": 43,
+                "json": {"some key": "some value"},
+                "jsonb": {"another key": [1, 2, "3"]}
+              } |]
+            `shouldRespondWith`
+              [json| {
+                "double": 3.1,
+                "varchar": "hello",
+                "boolean": true,
+                "date": "2019-01-01",
+                "money": "$0.00",
+                "enum": "foo",
+                "arr": ["a", "b", "c"],
+                "integer": 43,
+                "json": {"some key": "some value"},
+                "jsonb": {"another key": [1, 2, "3"]}
+              } |]
+              { matchHeaders = [matchContentTypeJson] }
+
+      it "accepts a variety of arguments with GET" $
+        -- without JSON / JSONB here, because passing those via query string is useless - they just become a "json string" all the time
```

Comment:
So passing JSON / JSONB via query string does not seem to work. At least if I try to pass something that should match an object, it seems like this is just returned as a json string.

Looking at the tests below that ("parses xxx JSON as JSON"), where this is explicitely tested for POST, I wonder: Is this expected behaviour for GET / query string (composing json objects in a query string is really ugly...) or is this to be considered a bug?

Reply:
The transformation from query params to sql params happens in Haskell code. I think it may be possible to fix this, but don't think is worth it. As you mentioned is ugly to deal with json(urlencoding) in the params.

I think users that wanted to pass json on GET params have been served well by `Prefer: params=single-object` http://postgrest.org/en/v7.0.0/api.html#calling-functions-with-a-single-json-parameter.

Reply:
@wolfgangwalther Now that I think about it maybe the shortcoming is in the ` second JSON.toJSON` in:

```haskell
rpcPrmsToJson = ProcessedJSON (JSON.encode $ mapFromListWithMultiple $ second JSON.toJSON <$> rpcQParams) (S.fromList $ fst <$> rpcQParams)
```

The fix might could be using something like this?(Not sure):

https://github.com/PostgREST/postgrest/blob/8e4687fb5301f9000c31c70f4fe22a87fe6d4d3f/src/PostgREST/Middleware.hs#L90-L95